runnable/extra-files/dexcept.h

```diff
@@ -0,0 +1,63 @@
+// D exceptions to by caught by C++
+
+#pragma once
+
+namespace D {
+
+struct string
+{
+    size_t length;
+    const char* ptr;
+};
+
+class Object;
+
+extern "C" {
+    string _d_toString(Object* o);
+}
+
+class Object
+{
+    void* __monitor;
+
+    virtual void _classinfo_data() = 0;
+
+protected:
+    // don't call directly, ABI mismatch

```

Comment:
I meant this test as a show case how easy it is now to mix exception handling between D and C++. Catching C++ exception in D needs a few changes to the front end (allowing C++ classes as catch type) that Walter added only very recently, so I guess it'll need some time until this is in LDC. Mirroring the D definitions with a C++ definition is just the same as declaring a C++ class in D.

Unfortunately, Walter chose a different ABI than the C++ compiler for x86 (http://dlang.org/spec/abi.html), especially `this` is passed in EAX, not ECX as for `__thiscall`. That's why mapping virtual functions doesn't work without making the class extern(C++) in D. I don't think this should be done to the D exceptions.

The actual virtual functions are just for documentation here, just exposing the useful wrappers should be ok, too.

Writing the binding function `_d_toString` on the D side has the big advantage that D knows both sides of the ABI, so you don't have to dig into assembly. It could be generalized to call any virtual function by specifying the offset in the vtbl, though.